### PR TITLE
[AutoDiff] Parameter indices data structure overhaul

### PR DESCRIPTION
**Pull request by rxwei**
### Introduce `AutoDiffIndexSubset`

This PR overhauls data structures used for parameter indices in the AutoDiff infrastructure in SIL.

Previously, we used `llvm::SmallBitVector` to represent differentiation parameter indices in both AST and SIL. It was not efficient, and most importantly there's no way to put this in an instruction without causing memory leaks.

This change replaces all uses of `llvm::SmallBitVector` in SIL AutoDiff code paths with a `ASTContext`-uniqued `AutoDiffIndexSubset *` where bits are stored as trailing objects.

`AutoDiffIndexSubset` does not have "parameter indices" in its name because it is not only designed for parameter indices, but also for result indices as we move to supporting multi-result differentiation. `AutoDiffIndexSubset` has set operations like `isSubsetOf`, `isSupersetOf`, and `contains`, but it also has a special _capacity_ property. All differentiable function's parameter indices data should store the number of parameters as capacity, so that the differentiation transform won't need special logic to check whether an index is out of range.

Another minor change is the module format layout of `SILDifferentiableAttr`. It used to store parameter indices as consecutive `bool` bits, but now stores numeric parameter indices directly for efficiency.

It will be necessary to refactor or eliminate `AutoDiffParameterIndices` to make use of `AutoDiffIndexSubset`. `AutoDiffParameterIndices` is at the AST level, so it is not in the scope for this PR.

Unblocks #23482. Partially resolves [TF-67](https://bugs.swift.org/browse/TF-67).

### CODE REVIEW

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

File: include/swift/AST/AutoDiff.h [hunk 1]
```diff
@@ -219,6 +222,252 @@ class AutoDiffParameterIndicesBuilder {
...
+  static unsigned getNumBitWordsNeededForCapacity(unsigned capacity) {
+    if (capacity == 0) return 0;
+    return capacity / numBitsPerBitWord + 1;
+  }
+  
+private:
+  /// The total capacity of the index subset, which is `1` less than the largest
+  /// index.
+  unsigned capacity;
+  /// The number of bit words in the index subset. in the index subset.
```
Comment (dan-zheng) [inline include/swift/AST/AutoDiff.h]:
```suggestion
  /// The number of bit words in the index subset.
```

File: include/swift/AST/DiagnosticsParse.def [hunk 2]
```diff
@@ -1535,7 +1535,9 @@ ERROR(sil_inst_autodiff_operand_list_expected_rbrace,PointsToFirstBadToken,
 ERROR(sil_inst_autodiff_num_operand_list_order_mismatch,PointsToFirstBadToken,
       "the number of operand lists does not match the order", ())
 ERROR(sil_inst_autodiff_expected_associated_function_kind_attr,PointsToFirstBadToken,
-      "expects an assoiacted function kind attribute, e.g. '[jvp]'", ())
+      "expected an assoiacted function kind attribute, e.g. '[jvp]'", ())
```
Comment (dan-zheng) [inline include/swift/AST/DiagnosticsParse.def]:
```suggestion
      "expected an associated function kind attribute, e.g. '[jvp]'", ())
```

**Review by dan-zheng — APPROVED** [review]
Hurray!

File: include/swift/AST/AutoDiff.h (hunk 1 above)
Comment (rxwei) [reply include/swift/AST/AutoDiff.h]: Good catch!

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow linux

**Pull request merged**